onnxruntime/test/providers/provider_test_utils.h

```diff
@@ -3,10 +3,17 @@
 
 #pragma once
 
+#include "core/providers/xnnpack/xnnpack_init.h"
+
 #include "test/providers/checkers.h"
 #include "test/providers/op_tester.h"
 #include "test/providers/model_tester.h"
 
+// XNNPACK_FP16_SUPPORTED scope is too big, so add USE_XNNPACK to avoid the FP16 tests enabled for other EPs
+#if defined(USE_XNNPACK) && defined(XNNPACK_FP16_SUPPORTED)
```

Comment:
I thought `XNNPACK_FP16_SUPPORTED` is defined if and only if `USE_XNNPACK`  is defined.
Is `XNNPACK_FP16_SUPPORTED`  independent?

Reply:
XNNPACK_F16_SUPPORTED only check the hardware environment.
Since it's defined in the XNNPACK package, it's strange to add USE_XNNPACK before it.
like MLAS_F16VEC_INTRINSICS_SUPPORTED, it doesn't dependent on USE_MLAS.

But a little surprise to me, if I don't add USE_XNNPACK, QNN or NNAPI EP will fail.
I think even one FP16 test is enabled for other EPs, they shouldn't be failed, they should give a warning that they don't support this FP16 kernel. Maybe some improvements should be added on these EPs.

Reply:
`USE_XNNPACK` a macro defiend in ONNXRUNTIME.  If `XNNPACK_F16_SUPPORTED` is defined inside Xnnpack package, `XNNPACK_F16_SUPPORTED` is undefined if `USE_XNNPACK` is undefined. As XNNPACK package woundn't be compiled if `USE_XNNPACK` is false.

Reply:
`QNN or NNAPI EP will fail` is quite weird as they are totally irrelevant of XNNPACK.

Reply:
NOT weird. Because  XNNPACK_F16_SUPPORTED scope is too wide. So I added USE_XNNPACK
It enables tests that NNAPI doesn't suppot.
e.g.
2024-10-09T02:59:51.5373194Z /mnt/vss/_work/1/s/onnxruntime/test/providers/base_tester.cc:830: registered execution providers: NnapiExecutionProvider
2024-10-09T02:59:51.5373411Z 
2024-10-09T02:59:51.5373646Z /mnt/vss/_work/1/s/onnxruntime/test/providers/base_tester.cc:317: Failure
2024-10-09T02:59:51.5373876Z Expected equality of these values:
2024-10-09T02:59:51.5374061Z   expect_result
2024-10-09T02:59:51.5374303Z     Which is: 4-byte object <00-00 00-00>
2024-10-09T02:59:51.5374511Z   ExpectResult::kExpectFailure
2024-10-09T02:59:51.5374762Z     Which is: 4-byte object <01-00 00-00>
2024-10-09T02:59:51.5375188Z Initialize failed but expected success: This is an invalid model. Type Error: Type 'tensor(float)' of input parameter (InsertedPrecisionFreeCast_X) of operator (NhwcFusedConv) in node (node1) is invalid.
2024-10-09T02:59:51.5375480Z Google Test trace:
2024-10-09T02:59:51.5375723Z /mnt/vss/_work/1/s/onnxruntime/test/providers/base_tester.cc:830: registered execution providers: NnapiExecutionProvider
2024-10-09T02:59:51.5375891Z 
2024-10-09T02:59:51.5376075Z [  FAILED  ] ConvFp16Test.Conv2D_Bias_Z_Relu (4 ms)

Reply:
> XNNPACK_F16_SUPPORTED scope is too wide.

I think the more correct fix is for xnnpack_init.h to have an `#if defined(USE_XNNPACK)` around its contents so it only defines things when the EP is enabled, as we're now including that header directly in any build.

Reply:
I think USE_XNNPACK is enough in test project.
XNNPACK_FP16_SUPPORTED is only used in XNNPACK provider project to decide the capability.
Thus, xnnpack_init.h will note be included by external projects.